src/tls/httpredirect.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include "httpredirect.h"
+
+#include <sys/socket.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+
+static char *
+read_line (FILE *stream,
+           char *buffer,
+           size_t sizeof_buffer,
+           size_t *offset)
+{
+  if (*offset >= sizeof_buffer)
+    return NULL;
+
+  char *line = fgets (buffer + *offset, sizeof_buffer - *offset, stream);
+
+  if (line == NULL)
+    return NULL;
+
+  /* Make sure we're terminated with \r\n or \n */
+  size_t line_length = strcspn (line, "\r\n");
+  const char *ending = line + line_length;
+  if (strcmp (ending, "\r\n") != 0 && strcmp (ending, "\n") != 0)
+    return NULL;
```

Comment:
Initially I thought "the `\r` check is completely insignificant here". Reading it again I realized it's not -- it makes sure that there is no `\r` in the *middle* of the line. Can you please amend the comment to say ".. and that there are no stray \r"?

Then again I wonder why we specifically handle `\r` but not any of the other ASCII control characters -- if they don't break the logic, why would `\r`? Sure, that could freak our clients when we throw their garbage back and them, but that's not our problem? Also, it feels like that doesn't belong into a generic function like `read_line()`.

IOW,  what would/could break if we changed this to
```C
if (line [strlen (line) - 1] != '\n')
    return NULL;
```

Reply:
`\r` is special here because it's not just whitespace.  the http spec says that you should send `\r\n` to terminate lines.  I tolerate bare '\n` as a courtesy to those who connect manually, but by spec, we actually don't need to.

Having `ending` set to the logical end of the data is important here: we need to trim the `\r` following `Host: foo`, for example, before we paste it together to make our redirection location and end up with `https://foo\r/path`.

Reply:
On the other hand, it's possible also to imagine that the field could theoretically contain other trailing whitespace, and we would then want to trim that as well, I guess.  Nobody ever actually sends that in practice, of course...